api/task_manager.cc

```diff
@@ -0,0 +1,202 @@
+    res.state = status.task_status.state;
+    res.is_abortable = status.task_status.is_abortable;
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.task_status.parent_id.to_sstring();
+    res.number = status.number.get_number(status.module);
+    res.shard = status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.task_status.progress_total;
+    res.progress_completed = status.task_status.progress_completed;
+    return res;
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return make_ready_future<json::json_return_type>(ctx.tm.local().get_modules());
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return do_with(req->query_parameters, [&ctx] (auto &params) {
+            return ctx.tm.map_reduce0([&params] (task_manager& tm) {
+                std::vector<tm::task_stats> res{};
+                if (params.contains("module")) {
+                    task_manager::module m = task_manager::get_module(params["module"]);
+                    if (tm.get_tasks().contains(m)) {
+                        for (auto& task : tm.get_tasks()[m]) {
+                            if (filter_tasks(task, params)) {
+                                tm::task_stats ts;
+                                ts.task_id = task->get_status().id.to_sstring();
+                                ts.state = task->get_status().state;
+                                res.push_back(ts);
+                            }
+                        }
+                    }
+                } else {
+                    for (auto& [module, tasks] : tm.get_tasks()) {
+                        for (auto& task : tasks) {
+                            if (filter_tasks(task, params)) {
+                                tm::task_stats ts;
+                                ts.task_id = task->get_status().id.to_sstring();
+                                ts.state = task->get_status().state;
+                                res.push_back(ts);
+                            }
+                        }
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
+        }).then([] (const std::vector<tm::task_stats>& res) {
+            return make_ready_future<json::json_return_type>(res);
+        });
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) {
+        utils::UUID id = utils::UUID{req->param["task_id"]};
+        return ctx.tm.map_reduce0([id] (task_manager& tm) -> status {
+            status s;
+            s.shard = this_shard_id();
+            s.number = tm.get_running_tasks_count();
```

Comment:
s.number is not supposed to be dynamic, but rather a running sequence number assigned when created, along with the task uuid. (see repair code).
It's for backward compatibility with async repair tasks.

Reply:
Oh, I misunderstood the purpose of number field. I will change that.